"Multiple VRFs and GUE DECAP in Default VRF" README.md (#3476)
openconfig/featureprofiles: feature/experimental/policy/policy_vrf_selection/otg_tests/multiple_vrfs_and_gue_decap/README.md

Reviewer: Please move this file into the folder already set up for `feature/policy_forwarding/policy_vrf_selection/otg_tests`.  ie:
https://github.com/openconfig/featureprofiles/tree/main/feature/policy_forwarding/policy_vrf_selection/otg_tests

Reviewer: The other vrf_selection tests should also be consolidated to a single folder if you don't mind moving them.

https://github.com/openconfig/featureprofiles/tree/main/feature/poliie move the tests from here: cy_forwarding/policy_vrf_selection/otg_tests

to here:
https://github.com/openconfig/featureprofiles/tree/main/feature/policy_forwarding/vrf_selection

Reviewer 2: Consolidated to the directory you suggested. PTAL at https://github.com/openconfig/featureprofiles/pull/4447 and close this PR

# RT-3.3: Multiple VRFs and GUE DECAP in Default VRF

## Summary

> This test ensures NOS is able to host multiple VRFs and at the sametime perform GUE tunnel DECAP in the default VRF

Reviewer 3: ```suggestion
This test ensures NOS is able to host multiple VRFs and at the same time perform GUE tunnel DECAP in the default VRF.
```

Reviewer: The latest commit still has the typo

## Procedure
Test environment setup

### Topology
Create the following connections:

> ```mermaid

Reviewer 3: (no action required) I had no clue you could do this, neat!

graph LR; 
subgraph ATE1 [ATE1]
    A1[Port1] 
    A2[Port2]
end

subgraph DUT1 [DUT1]
    B1[Port1]
    B2[Port2]
end

subgraph ATE2 [ATE2]
    C1[Port1] 
    C2[Port2]
end

A1 <-- IBGP(ASN100) --> B1; 
B2 <-- EBGP(ASN100:ASN200) --> C1;
A2 <-- EBGP(ASN100:ASN200) --> C2;
```

### Advertisements:

	ATE1:Port1 advertises following prefixes over IBGP to DUT:Port1
		- IPv4Prefix1/24 IPv6Prefix1/64
		- IPv4Prefix2/24 IPv6Prefix2/64
		- IPv4Prefix3/24 IPv6Prefix3/64
		- IPv4Prefix4/24 IPv6Prefix4/64
		- IPv4Prefix5/24 IPv6Prefix5/64
	
	ATE2:Port1 advertieses following prefixes to DUT:Port2 over EBGP
		- IPv4Prefix6/24 IPv6Prefix6/64
		- IPv4Prefix7/24 IPv6Prefix7/64
		- IPv4Prefix8/24 IPv6Prefix8/64
		- IPv4Prefix9/24 IPv6Prefix9/64
		- IPv4Prefix10/24 IPv6Prefix10/64

> 	DUT:Port2 advertises following tunnel endpoint prefixe to ATE2:Port1 over EBGP. 

Reviewer 3: ```suggestion
	DUT:Port2 advertises following tunnel endpoint prefixes to ATE2:Port1 over EBGP. 
```

		- IPv4Prefix12/28

	ATE1:Port2 advertises following tunnel endpoint prefix to ATE2:Port2
		- IPv4Prefix11/28

### Flows:

	From ATE1:Port1 to ATE2:Port1
		- IPv4Prefix1/24 to IPv4Prefix6/24 [DSCP:BE1]
		- IPv6Prefix1/64 to IPv6Prefix6/64 [DSCP:BE1]

		- IPv4Prefix2/24 to IPv4Prefix7/24 [DSCP:AF1]
		- IPv6Prefix2/64 to IPv6Prefix7/64 [DSCP:AF1]

		- IPv4Prefix3/24 to IPv4Prefix8/24 [DSCP:AF2]
		- IPv6Prefix3/64 to IPv6Prefix8/64 [DSCP:AF2]

		- IPv4Prefix4/24 to IPv4Prefix9/24 [DSCP:AF3]
		- IPv6Prefix4/64 to IPv6Prefix9/64 [DSCP:AF3]

		- IPv4Prefix5/24 to IPv4Prefix10/24 [DSCP:AF4]
		- IPv6Prefix5/64 to IPv6Prefix10/64 [DSCP:AF4]

	Flows from ATE2:Port1 to ATE1:Port1
		- IPv4Prefix6/24 to IPv4Prefix1/24 [DSCP:BE1]
		- IPv6Prefix6/64 to IPv6Prefix1/64 [DSCP:BE1]

		- IPv4Prefix7/24 to IPv4Prefix2/24 [DSCP:AF1]
		- IPv6Prefix7/64 to IPv6Prefix2/64 [DSCP:AF1]

		- IPv4Prefix8/24 to IPv4Prefix3/24 [DSCP:AF2]
		- IPv6Prefix8/64 to IPv6Prefix3/64 [DSCP:AF2]

		- IPv4Prefix9/24 to IPv4Prefix4/24 [DSCP:AF3]
		- IPv6Prefix9/64 to IPv6Prefix4/64 [DSCP:AF3]

		- IPv4Prefix10/24 to IPv4Prefix5/24 [DSCP:AF4]
		- IPv6Prefix10/64 to IPv6Prefix5/64 [DSCP:AF4]

>   - DUT has 2 VRFs, Default and Non-Default. Ensure leaking routes from the default to the non-default VRFs so prefixes learnt over IBGP and EBGP are available in the tables of bothe VRFs.

Reviewer 3: ```suggestion
  - DUT has 2 VRFs, Default and Non-Default. Ensure leaking routes from the default to the non-default VRFs so prefixes learnt over IBGP and EBGP are available in the tables of both VRFs.
```

>   - DUT has 2 VRFs, Default and Non-Default. Ensure leaking routes from the default to the non-default VRFs so prefixes learnt over IBGP and EBGP are available in the tables of bothe VRFs.

Reviewer 3: Maybe just worth clarifying (AIUI):

 * DUT port 1 and port 2 are assigned to the `DEFAULT` VRF.
 * BGP session between ATE1 and DUT1 is in the `DEFAULT` VRF.
 * BGP session between ATE2 and DUT2 is in the `DEFAULT` VRF.

So there are no routes learnt into the `Non-Default` VRF other than by importing them from the `DEFAULT` VRF. Right?

Reviewer 2: I have added clarity to the verbiage here.

  - DUT:PORT1 and DUT:Port2 are in the Default VRF.
  - Ensure IPv4Prefix12/28 is configured on the Loopback0 interface of DUT which is also in the Default VRF<br><br><br>

> **RT-3.3.1: [Baseline] Traffic flow between ATE1:Port1 and ATE2:Port1 via DUT's Default VRF**

Reviewer 3: You can make these a new heading to stand out further with `####`.

Reviewer 2: Done

  * Start above mentioned flows between ATE1:Port1 and ATE2:Port1.

>   * Ensure 100% success<br><br><br>

Reviewer 3: Maybe just clarify here -- 100% success means zero packet loss, I assume.

Reviewer 2: Done

> **RT-3.3.2: BE1 traffic from ATE1:Port1 to ATE2:Port1 simulated to be GUE Encaped and sent to the DUT's Default VRF by ATE2:Port1**

Reviewer 3: ```suggestion
#### RT-3.3.2: BE1 traffic from ATE1:Port1 to ATE2:Port1 simulated to be GUE Encapped and sent to the DUT's Default VRF by ATE2:Port1
```

Reviewer 2: Done

>   * ATE2:Port1 sends "Prefix1 to Prefix6" IPv4 and IPv6 traffic encaped inside tunnel destination IPv4Prefix12

Reviewer 3: To be clear here -- ATE:Port1 is sending an IPv4-in-UDP-in-IPv4 packet and then subsequently an IPv6-in-UDP-in-IPv4 packet?

Reviewer 2: Updated

  * ATE2:Port1 sends the following IPv4 and IPv6 flows as is
    * Prefix6/24 to IPv4Prefix1/24 [DSCP:BE1]
    * Prefix7/24 to IPv4Prefix2/24 [DSCP:AF1]
    * Prefix8/24 to IPv4Prefix3/24 [DSCP:AF2]
    * Prefix9/24 to IPv4Prefix4/24 [DSCP:AF3]
    * Prefix10/24 to IPv4Prefix5/24 [DSCP:AF4]
  * ATE1:Port1 continues sending the folllowing IPv4 and IPv6 traffic
    * Prefix2 to Prefix7 with DSCP:AF1
    * Prefix3 to Prefix8 with DSCP:AF2
    * Prefix4 to Prefix9 with DSCP:AF3
    * Prefix5 to Prefix10 with DSCP:AF4

>   * DUT:Port2 receives the tunneled traffic, decaps it, does a LPM lookup on the destination prefix (Prefix6) and routes it back to ATE2:Port1

Reviewer 3: What tells `DUT` to do this decapsulation? I'm not sure that it's mentioned anywhere in the README that the DUT is expected to be told that this is the case.

Reviewer 2: Clarified in Implementation section

  * All the traffic flows MUST show 100% success.
  * Streamed data on the number of packets decaped by the Tunnel endpoint "IPv4Prefix12" must match the number of tunnel encaped packets sent.<br><br><br>
  
   

> **RT-3.3.3 to RT-3.3.6: ATE2:Port1 simulates different traffic classes (BE1 to AF4) of FLOWS from ATE1:Port1 being GUE tunneled inside IPv4Prefix12 and sent to DUT:Port2**

Reviewer 3: ```suggestion
#### RT-3.3.3 to RT-3.3.6: ATE2:Port1 simulates different traffic classes (BE1 to AF4) of FLOWS from ATE1:Port1 being GUE tunneled inside IPv4Prefix12 and sent to DUT:Port2
```

Reviewer 2: Done

  * Follow each step in RT-3.3.2 for the following subtest cases
    * RT-3.3.3: BE1 and AF1 traffic from ATE1:Port1 to ATE2:Port1 simulated to be GUE Encaped and sent to the DUT's Default VRF by ATE2:Port1

>       * Since "Prefix1 to Prefix6" and "Prefix2 to Prefix7" IPv4 and IPv6 flows are sent inside the tunnel, ATE1:Port1 musn't send these flows

Reviewer 3: Can you clarify what is being said here? We have a flow going from prefix 1 --> prefix 7 (I presume src and dst addresses...), the DUT is receiving prefix 7 via eBGP from ATE-2, why is this traffic dropped? Doesn't the DUT forward the packets (even if it has no decapsulation configuration -- which this doesn't specify)?

I think I have a similar misunderstanding about the cases below.

> # RT-3.3: Multiple VRFs and GUE DECAP in Default VRF

Reviewer: Recommend using "PF" which is what is usually used for the feature/policy-forwarding tests.  Maybe make it 2.x, to map it as  policy_forwarding/vrf_selection?  (I realize the other vrf_selection tests used RT, but let's make the change here)

```suggestion
# PF-2.3: Multiple VRFs and GUE DECAP in Default VRF
```
